src/main/groovy/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.groovy

```diff
@@ -168,52 +169,63 @@ class GitRepository implements ScmRepository {
                 repository.branch.current.name
         )
     }
-
+    
     TagsOnCommit latestTags(Pattern pattern) {
-        return latestTagsInternal(pattern, null, true)
+      return latestTagsInternal(pattern, null, true)
     }
-
+  
     TagsOnCommit latestTags(Pattern pattern, String sinceCommit) {
         return latestTagsInternal(pattern, sinceCommit, false)
     }
-
+    
     private TagsOnCommit latestTagsInternal(Pattern pattern, String maybeSinceCommit, boolean inclusive) {
-        if (!hasCommits()) {
-            return new TagsOnCommit(null, [], false)
-        }
-
-        Map<String, List<String>> allTags = tagsMatching(pattern)
-
-        ObjectId headId = repository.repository.jgit.repository.resolve(Constants.HEAD)
-
-        ObjectId startingCommit;
-        if (maybeSinceCommit != null) {
-            startingCommit = ObjectId.fromString(maybeSinceCommit)
-        } else {
-            startingCommit = headId
-        }
-
-        RevWalk walk = walker(startingCommit)
-        if (!inclusive) {
-            walk.next()
-        }
-
-        List tagsList = null
-
-        RevCommit commit
-        for (commit = walk.next(); commit != null; commit = walk.next()) {
-            tagsList = allTags[commit.id.name()]
-            if (tagsList) {
-                break
-            }
-        }
-        walk.dispose()
-
-        if (commit == null) {
-            return new TagsOnCommit(null, [], false)
-        }
-
-        return new TagsOnCommit(commit.id.name(), tagsList, Objects.equals(commit.id, headId))
+      List<TagsOnCommit> allTaggedCommits = allTaggedCommits(pattern, maybeSinceCommit, inclusive, true)
+      TagsOnCommit taggedCommit = null
+      if (allTaggedCommits.size() == 0) {
+        taggedCommit = new TagsOnCommit(null, [], false)
+      } else {
+        taggedCommit = allTaggedCommits.get(0)
+      }
+      return taggedCommit
+    }
+    
+    List<TagsOnCommit> allTaggedCommits(Pattern pattern, String maybeSinceCommit, boolean inclusive, boolean stopOnFirstTag) {
```

Comment:
this method should not be public, as it has a lot of arguments, look at `latestTags` API: it has two simple public methods with the `latestTagsInternal`. same should go here: this method name should not be `allTaggedCommits`, as it has flag to return only one, not all. so instead it would be `taggedCommits` with public methods like `allTaggedComits(Pattern)` and `allTaggedCommits(Pattern, SinceCommit)` (if necessary)